File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-15.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0, right = nums.size() - 1;
        int middle{};
        // [5,6,1,2,3,4] & target = 6

        while (left <= right){
            middle = std::floor((static_cast<double>(left + right) / 2));

            // case where middle is equal to target
            if (nums[middle] == target) return middle;

            // left half is sorted
            if (nums[left] <= nums[middle]){
                // target exists within left -> middle
                if (target < nums[middle] && target >= nums[left]){
                    right = middle - 1;
                } else {
                    // target is not between left -> middle
                    left = middle + 1;
                }
            // left half is NOT sorted, but right side IS DEFINITELY sorted
            } else {
                if (target > nums[middle] && target <= nums[right]){
                    left = middle + 1;
                } else {
                    right = middle - 1;
                }
            }

        }
        return -1;
    }
};

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-15.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        
        left, right = 0, len(nums) - 1
        res = nums[0]

        while left <= right:

            # [4, 5, 6, 7, 1, 2, 3]

            if nums[left] < nums[right]: 
                res = min(res, nums[left])
                return res

            middle = (left + right) // 2

            res = min(res, nums[middle])

            if ( nums[left] <= nums[middle] ):
                left = middle + 1 
            else:
                right = middle - 1
        
        return res
        
